community/rfcs/techniques/design/20200306-weight-clustering-design.md

```diff
@@ -0,0 +1,144 @@
+
+The code snippets below illustrate common usage patterns of the clustering API. For the sake of brevity, the clustering examples presented here only deal with the 1st step of the above workflow.
+
+A more comprehensive end-to-end example on how to use the API to cluster and save a model is provided in the examples directory, under *clustering*.
+
+### Cluster an Entire Model
+
+When passing an entire model to `cluster_weights()`, the tool will determine which layers are clusterable and will cluster all of them &mdash; other layers will be ignored and left as they are.
+
+```python
+clustering_params = {
+    'number_of_clusters': 8,
+    'cluster_centroids_init': 'linear'
+}
+
+# Clustering a sequential model
+model = keras.Sequential([
+    layers.Dense(10, input_shape=(10,)),
+    layers.Dense(10),
+])
+
+clustered_model = cluster.cluster_weights(model, **clustering_params)
+clustered_model.build(input_shape=(1, 10))
+```
+
+```python
+# Clustering a functional model
+i1 = keras.Input(shape=(10,))
+i2 = keras.Input(shape=(10,))
+x1 = layers.Dense(10)(i1)
+x2 = layers.Dense(10)(i2)
+outputs = layers.Add()([x1, x2])
+
+model = keras.Model(inputs=[i1, i2], outputs=outputs)
+
+clustered_model = cluster.cluster_weights(model, **clustering_params)
+```
+
+### Cluster Some Layers
+
+Individual layers within a model can be clustered selectively by invoking `cluster_weights()` on a layer object, as follows:
+
+```python
+clustered_model = keras.Sequential()
+clustered_model.add(cluster.cluster_weights(keras.layers.Dense(10), **clustering_params))
+clustered_model.add(keras.layers.Dense(10))
+```
+In the above example model the weights of the 1st `Dense` layer will be clustered, whereas the weights of the 2nd layer will not be clustered.
+
+### Export Clustered Model
+
+```python
+clustered_model = cluster_weights(model, **clustering_params)
+keras.models.save_model(clustered_model, keras_file)
+```
+
+### Import Clustered Model
+
+A clustered Keras model needs to be loaded within a `cluster_scope` in order to be successfully deserialized:
+
+```python
+with cluster_scope():
+  loaded_model = keras.models.load_model(keras_file)
+```
+
+## Future Improvements
+
+This is a list of possible future improvements and features that are currently under consideration:
+
+* Support for custom clusterable layers &mdash; implemented already to some extent, but the API needs further refinement.
+* Sparsity-aware clustering, to ensure that the sparsity of pruned models is not destroyed during the clustering and fine-tuning process.
+* Clustering schedule and callback, to provide more advanced training-time control of the clustering behaviour and progressive layer clustering throughout training.
+* Improved training / fine-tuning by updating cluster-to-weight associations during the fine-tuning phase. This one is likely to be a good use case for having support for clustering schedule in the API.
+* More cluster centroid initialization methods, e.g. k-means
+* ...
```

Comment:
Please add a comparisons with existing TFMOT tools section. See the pruning RFC for an example.